### PR TITLE
Find the project ts server path, else fallback to bundled ts server path

### LINKED ISSUE #19: Use the typescript server from the current project
Currently LSP-volar has a hard-coded the path to the tsserver:

https://github.com/sublimelsp/LSP-volar/blob/2d01fd3504d58c52ac2bdf08c0be00a329f64129/LSP-volar.sublime-settings#L4-L7

LSP-volar assumes that the node_modules will be at the root of the opened folder, which is not always the case.

Proposal,
find the tsserver path, even when the node_modules is not at the root of the project,
and if it cannot find one fallback to a bundled tsserver path. (which requires typescript as a new dependency in the [package.json](https://github.com/sublimelsp/LSP-volar/blob/master/server/package.json#L1) file)

### PR DESCRIPTION
**Pull request by predragnikolic**
closes #19

### CODE REVIEW

File: plugin.py [hunk 1]
```diff
@@ -14,3 +19,25 @@ class LspVolarPlugin(NpmClientHandler):
...
+        cls,
+        window: sublime.Window,
+        initiating_view: Optional[sublime.View] = None,
+        workspace_folders: Optional[List[WorkspaceFolder]] = None,
+        configuration: Optional[ClientConfig] = None
+    ):
+        if not workspace_folders or not configuration:
+            return
+        ts_path = ''
+        resolve_module_script = os.path.join(cls._server_directory_path(), 'resolve_module.js')
```
Comment (predragnikolic) [inline plugin.py:34]: I do not like that I had to use a private method to get the` _server_directory_path`.
Any suggestions?

File: plugin.py [hunk 2]
```diff
@@ -14,3 +19,25 @@ class LspVolarPlugin(NpmClientHandler):
     package_name = __package__
     server_directory = 'server'
     server_binary_path = os.path.join(server_directory, 'node_modules', '@volar', 'server', 'out', 'index.js')
+
+    @classmethod
+    def is_allowed_to_start(
```
Comment (predragnikolic) [inline plugin.py:24]: I do not like that I had to use is_allowed_to_start.

I think that  `def on_client_configuration_ready(cls, configuration: Dict) -> None:` is a more appropriate hook.
https://github.com/sublimelsp/lsp_utils/blob/master/st3/lsp_utils/generic_client_handler.py#L181

But if I use `on_client_configuration_ready` I face two problems:
1. I have no access to workspace_folders. Alternative solution for that is to use `sublime.active_window().folders()`
2. `cls._server_directory_path()` will return me an empty string when called from `on_client_configuration_ready`. Alternative solution is to use `is_allowed_to_start` and modify the client configuration there.

File: plugin.py [hunk 3]
```diff
@@ -14,3 +19,25 @@ class LspVolarPlugin(NpmClientHandler):
...
+        initiating_view: Optional[sublime.View] = None,
+        workspace_folders: Optional[List[WorkspaceFolder]] = None,
+        configuration: Optional[ClientConfig] = None
+    ):
+        if not workspace_folders or not configuration:
+            return
+        ts_path = ''
+        resolve_module_script = os.path.join(cls._server_directory_path(), 'resolve_module.js')
+        first_folder = workspace_folders[0].path
+        proc = subprocess.Popen(['node', resolve_module_script, first_folder, 'typescript/lib/tsserverlibrary.js'], stdout=subprocess.PIPE)
```
Comment (rchl) [inline plugin.py:36]: Instead of `node` use `cls._node_bin()`.

Comment (rchl) [reply plugin.py:36]: Would it work to use higher-level API `subprocess.check_output` (with `universal_newlines=True`)?

File: plugin.py [hunk 4]
```diff
@@ -14,3 +19,25 @@ class LspVolarPlugin(NpmClientHandler):
...
+        if proc.stdout:
+            ts_path = proc.stdout.read().decode('utf-8', 'ignore') # workspace ts path
+        if not ts_path:
+            ts_path = os.path.join(cls._server_directory_path(), 'node_modules', 'typescript', 'lib', 'tsserverlibrary.js') # bundled ts path
+        configuration.init_options.set('typescript', {
+            'serverPath': ts_path
+        })
```
Comment (rchl) [inline plugin.py:43]: Same but simpler:
```suggestion
        configuration.init_options.set('typescript.serverPath', ts_path)
```

**Review by rchl — CHANGES_REQUESTED** [review]

File: plugin.py (hunk 1 above)
Comment (rchl) [reply plugin.py:34]: Same for `_node_bin` that I've suggested to use elsewhere.

You could use the public method `cls.get_server` to get the server instead but either way is fine probably.

File: plugin.py (hunk 3 above)
Comment (predragnikolic) [reply plugin.py]: Thanks.
Yes, it works.

File: plugin.py (hunk 1 above)
Comment (predragnikolic) [reply plugin.py:33]: `cls.get_serve` could return `None`. 
I do not want to add an additional if statement. 
I will stick with `_server_directory_path`   :)

**Review by rchl — APPROVED** [review]
I suppose you could also respect `typescript.serverPath` if explicitly set.

**Comment by rchl** [comment]
General lint notes:
 - recommended to have `"ensure_newline_at_eof_on_save": true,`
 - lines too long (over 120 chars)

**Pull request merged**